File: scala/src/P07Monad.scala
object P07Monad {
  def main(args: Array[String]): Unit = {
    val some4 = Some(4)
    val some2 = Some(2)

    val fmapResult = OptionFunctor fmap(f, some4)
    println(fmapResult)

    val applyResult = OptionApplicative apply(fmapResult, some2)
    assert(applyResult == Some(Some(2)))
    assert((applyResult flatMap { x => x }) == Some(2))

    val bindResult = OptionMonad bind(some2, fmapResult.get)
    assert(bindResult == Some(2))
  }

  val f: Int => Int => Option[Int] = x =>
    y => if (y == 0) None else Some(x / y)
}

trait Monad[M[_]] extends Applicative[M] {
  def bind[A, B](ma: M[A], f: A => M[B]): M[B]
}

object OptionMonad extends Monad[Option] {
  override def fmap[A, B](f: A => B, fa: Option[A]): Option[B] = {
    fa match {
      case None => None
      case Some(a) => Some(f(a))
    }
  }

  override def apply[A, B](f: Option[A => B], fa: Option[A]): Option[B] = {
    fa match {
      case None => None
      case Some(a) => f match {
        case None => None
        case Some(g) => Some(g(a))
      }
    }
  }

  override def bind[A, B](ma: Option[A], f: A => Option[B]): Option[B] = {
    ma match {
      case None => None
      case Some(a) => f(a)
    }
  }
}

File: haskell/P07Monad.hs
import Data.Typeable
import Test.HUnit.Base

main :: IO ()
main = do
  show (typeOf fmapResult) @?= "Maybe (Int -> Maybe Int)"
  applyResult @?= Just (Just 2)
  bindResult @?= Just 2

some4 = Just 4
some2 = Just 2

fmapResult = fmap' f some4

applyResult = apply' fmapResult some2

unwrapped :: Maybe(Int -> Maybe Int) -> (Int -> Maybe Int)
unwrapped Nothing = error "nothing to unwrap"
unwrapped (Just a) = a

bindResult = bind' some2 (unwrapped fmapResult)

f :: Int -> Int -> Maybe Int
f x y = if y == 0 then Nothing else Just (x `div` y)

class Functor' f where
  fmap' :: (a -> b) -> f a -> f b

class Functor' f => Applicative' f where
  apply' :: f (a -> b) -> f a -> f b

class Applicative' m => Monad' m where
   bind' :: m a -> (a -> m b) -> m b

instance Functor' Maybe where
  fmap' _ Nothing = Nothing
  fmap' f (Just a) = Just (f a)

instance Applicative' Maybe where
  apply' _ Nothing = Nothing
  apply' Nothing _ = Nothing
  apply' (Just g) (Just a) = Just (g a)

instance Monad' Maybe where
  bind' Nothing _ = Nothing
  bind' (Just a) f = f a